### PR TITLE
Add case of cgroup vhost

### PR DESCRIPTION
**Pull request by chloerh**
- VIRT-300725 - [cgroup][vhost] Check the vhost* thread's cpu affinity with emulatorpin setting

Test result:
PASS 1-type_specific.io-github-autotest-libvirt.virtual_network.check_vhost_cpu_affinity_with_emulatorpin

### CODE REVIEW

**Review by yalzhang — APPROVED** [review]
It can cover the bug succesfully. Run with old kernel, fail, while run with updated kernel, pass.

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg [hunk 1]
```diff
@@ -0,0 +1,8 @@
+- virtual_network.check_vhost_cpu_affinity_with_emulatorpin:
```
Comment (dzhengfy) [inline libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg:1]: Currently there are many files or folders under virtual_network. 
I think we'd better make an organized structure. 
@chloerh @yalzhang How do you think to put this new file?

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    type = check_vhost_cpu_affinity_with_emulatorpin
+    start_vm = no
+    timeout = 240
+    vcpu_set = 0
+    epin_set = 1
```
Comment (dzhengfy) [inline libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg:6]: This hardcoded cpu id 1 is at a little risky. Sometimes the cpu 1 might be offlined or non usable.  I suggest we dynamically choose another online cpu id.

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        if epin['*'] != epin_set:
+            test.fail(f'Incorrect cpu affinity (emulatorpin): {epin["*"]},'
+                      f'Should be {epin_set}.')
+
+        pid = process.run(
+            "ps -eL|awk '/vhost/{print $2}'", shell=True).stdout_text.strip()
+
+        taskset_output = process.run(
+            f'taskset -cap {pid}', shell=True).stdout_text
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:58]: It is better to firstly check the availability of `taskset` and install it if no.

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+def run(test, params, env):
+    """
+    Check the vhost* thread's cpu affinity with emulatorpin setting
+    """
+    vm_name = params.get('main_vm')
+    vm = env.get_vm(vm_name)
+    vcpu_set = params.get('vcpu_set')
+    epin_set = params.get('epin_set')
+    epin_reset = random.randint(int(epin_set), 10)
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:25]: This is at risk. Please choose from online cpus. @chloerh 
Do you need to consider different numa nodes？ @yalzhang

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            f'taskset -cap {pid}', shell=True).stdout_text
+        search_pattern = "pid \d+'s current affinity list:\s(\S+)"
+        search_result = re.search(search_pattern, taskset_output)
+        if not search_result or search_result.group(1) != epin_set:
+            test.fail(f'Expect current affinity to be {epin_set}, '
+                      f'not {search_result.group(1)}.')
+
+        # Update the emulator pin to other cpu,
+        # and check the vhost thread's cpu affinity
+        virsh.emulatorpin(vm_name, epin_reset, **VIRSH_ARGS)
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:67]: Line 67 - 80 seems many duplcation with line 42-58. I think your codes are able to be more beatiful if removing those duplication.

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg (hunk 2 above)
Comment (chloerh) [reply libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg:6]: @yalzhang Could you help confirm? Thanks

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg (hunk 1 above)
Comment (chloerh) [reply libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg:1]: Sure, I'm ok either way.

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 3 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:58]: ok

Comment (chloerh) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:58]: Actually this is on host, I just need to check its existence?

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 4 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:25]: Sure, seems the polarion case needs an update too.

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg (hunk 2 above)
Comment (yalzhang) [reply libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg:6]: Yes, we can choose from the online cpus. The test point of this case is that the vhost thread is pinned to the same cpu as the emulator.

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg (hunk 1 above)
Comment (yalzhang) [reply libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg:1]: Maybe we can create a folder called "driver" under the "virtual network" folder. There should be several cases about dirver later like multiqueue, we can put them together here. What do you think?

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 3 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:58]: No need. This command is provided by util-linux-core, which is needed by systemd. So I think it must exists on the host.
# yum remove util-linux-core
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use "rhc" or "subscription-manager" to register.

Error: 
 Problem: The operation would result in removing the following protected packages: systemd
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 4 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:25]: No, no need to consider the numa nodes. The check point is that the vhost thread will always pinned to the same cpu as the emulator. So in this step, we just update the affinity of th emulator to check it the vhost thread affinity will change accordingly.

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 3 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py]: ok

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 4 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py]: Done

File: libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py (hunk 5 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.py:98]: Done

**Comment by chloerh** [comment]
All resolved except the one of putting file to another folder, will update after confirmation.
Test result:
` (1/1) type_specific.io-github-autotest-libvirt.virtual_network.check_vhost_cpu_affinity_with_emulatorpin: PASS (7.57 s)`

File: libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg (hunk 1 above)
Comment (chloerh) [reply libvirt/tests/cfg/virtual_network/check_vhost_cpu_affinity_with_emulatorpin.cfg]: Done

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**